### ISSUE TITLE
Why are .beam files included into the repo?

### ISSUE THREAD

**Issue by dolfinus** [erlang/otp#4966]
I've studied the project and its structure when I found commits like this one:
https://github.com/erlang/otp/pull/4515

It looks like compiled .beam files were added into the repo:
https://github.com/erlang/otp/tree/master/erts/preloaded/ebin

As well as their source code files:
https://github.com/erlang/otp/tree/master/erts/preloaded/src

Why .beam files were added into the repo in the first place?

I mean git is a source code tracking system. It is designed for working with the source code, making branches which can be merged into other ones. making commits which can be used for tracking the history of each changed line and so on.
Adding some binaries (like image files) is okay because although they cannot be easily merged, you can still make a diff (most of diff viewers support that), see which files have been changed in a certain commit/branch, and also you can still move back in forward through the history to get a snapshot of the code base in any moment of time.

But:
1. .beam files are already based on the other files in the repo, like .erl, so they can be compiled any time you need
2. They are not updated in the same commit but some time later
3. They could create merge conflicts which is hard to resolve because there is no way to partially merge them, you should always rebase current commit/branch and compile them again
4. They are causing the increase of the diff size so its harder to read it
5. Some compilers which are making native binaries/executables/images could contain some parts depending on an operating system or CPU architecture. They could be in a binary format because they are not based on some source code in the repo. But in most cases they are just ASM/C/C++/Rust files which are compiled into binary files on the release stage.
6. BEAM is a virtual machine and .beam files cannot depend on the OS or CPU architecture at all. So there is nothing unique in this files to be kept in the repo.
7. I haven't seen any any compiled files) in repos of other VM-based languages (.jar files for Java).

So I don't see any need to track these files in the repo. Why do you use such an approach?

**Comment by lhoguin** [comment]
Erlang files are compiled using a compiler written in Erlang, so you need to have some beam files in order to bootstrap the build.

> .beam files are already based on the other files in the repo, like .erl, so they can be compiled any time you need

You need the .beam files to compile those .erl files. :-)

**Comment by bjorng** [comment]
As @lhoguin said.

A long time ago (before we used Git) we did not commit the preloaded modules and the primary bootstrap to the repository, but depended on every developer having already downloaded a previous release of Erlang. That had its own set of drawbacks. For one, new language features added to the compiler could not be used within OTP itself in the same release as they were introduced in. Having tried it both ways, I would say that the advantages of the current approach outweigh the disadvantages.
